### PR TITLE
Flash simulator stub

### PR DESCRIPTION
**Pull request by de-nordic**
The PR contains two commits:
 -  implementing the flash simulator stub, a device that always read erase_value and does nothing on write/erase; the device still respects offset, size and other parameters of flash simulator as set in a DTS;
  -  tests

### CODE REVIEW

**Comment by de-nordic** [comment]
If we decide with this approach, the other:
 - https://github.com/nrfconnect/sdk-zephyr/pull/623
 - https://github.com/zephyrproject-rtos/zephyr/pull/38524
could be dismissed.
The advantage of this approach is that it only requires NRF change, no need to update sdk-zephyr.

**Comment by tejlmand** [comment]
I like the approach, but I think we should avoid using the name `simulator` and name it according to it's purpose.

I think @carlescufi proposed something like `ram_flash()`, and then make it possible to use on both sides, i.e. the app side and in mcuboot, so that we completely can avoid referring to flash simulator.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (tejlmand) [inline drivers/flash/flash_simulator_stub/Kconfig:6]: following the comment here https://github.com/nrfconnect/sdk-nrf/pull/5620#issuecomment-919983174
I think `FLASH_RAM` might be a better naming.

**Comment by de-nordic** [comment]
> 
> 
> I like the approach, but I think we should avoid using the name `simulator` and name it according to it's purpose.
> 
> I think @carlescufi proposed something like `ram_flash()`, and then make it possible to use on both sides, i.e. the app side and in mcuboot, so that we completely can avoid referring to flash simulator.

I can rename it to whatever you want it to be, just know that, at least for now, it will always replace "zephyr,sim-flash" driver.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (tejlmand) [inline drivers/flash/flash_simulator_stub/Kconfig:7]: Is there a good reason why this symbol is having a prompt ?

Shouldn't it be prompt less and then internal Kconfig can select this symbol when mcuboot is enabled for thingy53.
(Other board may do similar when they start supporting dual image updates)

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (tejlmand) [inline drivers/flash/flash_simulator_stub/Kconfig:8]: should we have this ?
Usually flash drivers are not depending on another driver not being selected.

Of course selecting multiple drivers might end in linking errors later, but if we test for `!FLASH_SIMULATOR` then you should probably also test for `!NORDIC_QSPI_NOR`, `!SPI_NOR` etc.
https://github.com/nrfconnect/sdk-zephyr/blob/a64e96d17cc741937a8f5ff01a824ef0df22f662/drivers/flash/Kconfig.nordic_qspi_nor#L4
https://github.com/nrfconnect/sdk-zephyr/blob/a64e96d17cc741937a8f5ff01a824ef0df22f662/drivers/flash/Kconfig.nor#L5

File: tests/drivers/flash_simulator_stub/prj.conf
Comment (tejlmand) [inline tests/drivers/flash_simulator_stub/prj.conf:10]: This would move to a `Kconfig` file with content:
```
config FLASH_RAM
    bool
    default y

config FLASH_RAM_UNALIGNED_READ
    bool
    default y
```

**Review by tejlmand — COMMENTED** [review]
some small comments

**Comment by tejlmand** [comment]
> Update: Missed the "and then make it possible to use on both sides". OK will do other PR with that proposal too.

I think the intention was that one driver could be used as either full ram flash mapping, or as a stub (address handling) as it does in current PR state.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (de-nordic) [reply drivers/flash/flash_simulator_stub/Kconfig:8]: Yes, but other drivers are not cross compatible with other devices. This is alternative driver for the same "zephyr,sim-flash" device.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (nvlsianpu) [reply drivers/flash/flash_simulator_stub/Kconfig:7]: It is no for mcuboot but for the DFU app. Mcuboot is using upstream flash_simulator

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (de-nordic) [reply drivers/flash/flash_simulator_stub/Kconfig:8]: Other driver do not collide with it.

Comment (nvlsianpu) [reply drivers/flash/flash_simulator_stub/Kconfig:8]: It uses same configuration as flash_simulator for make this patch easier to be done.

**Comment by pdunaj** [comment]
Release is blocked as we cannot progress without this change.

**Are you sure this is a right moment to complain that code that should do nothing except return success does not have battery of tests or is named foo instead of loo.**

I recommend to keep it as simple as possible. This should not be 13 files that are hard to review. Let's throw away everything that is non-essential. Quickly and carefully check these few lines that remain. Let's name it whatever please the reviewer (give 5 minutes to come up with another name). And get back to testing the DFU.

It's not perfect code/name/place but I would rather we waste time on testing the thing depending on this driver then loose time here.

@tejlmand @carlescufi @shantha-14 

There are apparently 3 PRs for this created already <facepalm>.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]
I see that It is flash stub device we were discussing before. Usage of this in DFU application as temporary hack is fine IMO. We should fix this after probably via making subsystems capable to serve configuration in which primary slot might be unreachable for the DFU application.

**Comment by tejlmand** [comment]
just a question, if we had followed the path layed out in this PR earlier, then this commit would not have been needed upstream, is that correct ? @sigvartmh  @nvlsianpu 
https://github.com/zephyrproject-rtos/zephyr/pull/37783/commits/2d51e73198aeafc654e7aee3f8d11cceb654f538

**Comment by nvlsianpu** [comment]
@tejlmand No this patch is just the stub flash device.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (nvlsianpu) [reply drivers/flash/flash_simulator_stub/Kconfig:6]: Isn't it the stub?

**Comment by de-nordic** [comment]
@nvlsianpu  Changed name to FLASH_NOP_DEVICE as requested.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (tejlmand) [reply drivers/flash/flash_simulator_stub/Kconfig]: > 
> It is no for mcuboot but for the DFU app. Mcuboot is using upstream flash_simulator

@nvlsianpu I assume that what mcuboot will be using will depend on what we choose to do in this commit (which is a fromlist):
https://github.com/nrfconnect/sdk-mcuboot/pull/175/commits/e67d3dc27be1bb8b9f36e607cb0cecf78e448d65

as well as the possibility in sdk-nrf to choose alternative Kconfig settings for mcuboot as a child image.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (tejlmand) [reply drivers/flash/flash_simulator_stub/Kconfig]: I was referring to our talks regarding the simulator naming not being optimal name for NCS users.
The name simulator might be confusing to them when the look at this.

So why not choose a better name now, and not having to rename later ?

Remember, simulator will pop up in Kconfig, conf files, file names in IDEs, etc.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (tejlmand) [reply drivers/flash/flash_simulator_stub/Kconfig]: yes, but the DFU is only enabled if mcuboot is enabled, afaik.

Meaning this symbol is not something a user / conf file should set, but should be automatically selected when mcuboot and DFU for dual image update is enabled.
And this is being enabled in the app.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (tejlmand) [reply drivers/flash/flash_simulator_stub/Kconfig]: fair enough.

**Comment by tejlmand** [comment]
> Release is blocked as we cannot progress without this change.
>
> Are you sure this is a right moment to complain that code that should do nothing except return success does not have battery of tests or is named foo instead of loo.

@pdunaj I don't think anyone was blocking this PR when you wrote this comment.
But we should always comment things we observe, remember that if things are unclear to a reviewer, then it will be even more unclear to a user of this and that will lead to a bad user-experience and/or increased tech support burden.

**Comment by greg-fer** [comment]
Please add entry to changelog or remove the related label if the entry is not needed. Thanks.

File: drivers/flash/CMakeLists.txt
Comment (MarekPieta) [inline drivers/flash/CMakeLists.txt:2]: 2021

File: drivers/flash/flash_nop_device/Kconfig
Comment (MarekPieta) [inline drivers/flash/flash_nop_device/Kconfig:3]: 2021

File: drivers/flash/flash_nop_device/Kconfig
Comment (MarekPieta) [inline drivers/flash/flash_nop_device/Kconfig:28]: FLASH_NOP_DEVICE

File: drivers/flash/flash_nop_device/flash_nop_device.c
Comment (MarekPieta) [inline drivers/flash/flash_nop_device/flash_nop_device.c:14]: Do we need to include random, stats and string ?

File: drivers/flash/flash_nop_device/flash_nop_device.c
Comment (MarekPieta) [inline drivers/flash/flash_nop_device/flash_nop_device.c:151]: Add missing CODEOWNER to the files, please?

File: tests/drivers/flash_nop_device/CMakeLists.txt
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/CMakeLists.txt:2]: 2021

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:21]: What do you mean in the comment ?

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:34]: We do not erase device here

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:69]: assert that we can write value of this size (compatible with FLASH_NOP_DEVICE_PROG_UNIT)

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:81]: flash read

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:95]: For returned values we could use `zassert_ok` (more readable)

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:103]: Should we use FLASH_NOP_DEVICE_PROG_UNIT instead of hardcoded value of `4` ?

Comment (MarekPieta) [reply tests/drivers/flash_nop_device/src/main.c:103]: Also should we check if `FLASH_NOP_DEVICE_BASE_OFFSET >= 4` ?
Otherwise we may see underflow here

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:159]: Should we use FLASH_NOP_DEVICE_PROG_UNIT  ?

Comment (MarekPieta) [reply tests/drivers/flash_nop_device/src/main.c:159]: And skip test if it's 1 ?

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:186]: Consider also checking write_block_size

File: drivers/flash/Kconfig
Comment (MarekPieta) [inline drivers/flash/Kconfig:7]: btw. FLASH appears now twice in Kconfig. Do we need a separate symbol in NCS ?

**Review by pdunaj — APPROVED** [review]
assuming small issues in dates and options are fixed
don't reviewing tests

**Review by nvlsianpu — APPROVED** [review]

File: drivers/flash/flash_nop_device/flash_nop_device.c
Comment (de-nordic) [reply drivers/flash/flash_nop_device/flash_nop_device.c]: strings is bringing memset, others have been removed.

File: tests/drivers/flash_nop_device/src/main.c
Comment (de-nordic) [reply tests/drivers/flash_nop_device/src/main.c:159]: The problem persists also in upstream flash_simulator device test. The test has basically been hardcoded, here and there, for the 4 byte alignment check to fail. Will not fix it in this PR.

File: drivers/flash/Kconfig
Comment (de-nordic) [reply drivers/flash/Kconfig:7]: I do not think so, might be wrong, but this way the FLASH, in NRF drivers, gets automatically visible when you select FLASH in zephyr.

**Review by sigvartmh — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by de-nordic** [comment]
> 
> 
> just a question, if we had followed the path layed out in this PR earlier, then this commit would not have been needed upstream, is that correct ? @sigvartmh @nvlsianpu
> [zephyrproject-rtos/zephyr@2d51e73](https://github.com/zephyrproject-rtos/zephyr/commit/2d51e73198aeafc654e7aee3f8d11cceb654f538)

The upstream PRs will be dropped. The mcumr will be later modified to support situation where there are DTS defined images that do not have devices and situations where there could be only one slot available for images with index  > 0.

**Comment by de-nordic** [comment]
> 
> 
> > Update: Missed the "and then make it possible to use on both sides". OK will do other PR with that proposal too.
> 
> I think the intention was that one driver could be used as either full ram flash mapping, or as a stub (address handling) as it does in current PR state.

We probably need to create, at some point, special "RAM window" driver that would be something separate from the "zephyr,sim-flash" driver.

File: tests/drivers/flash_simulator_stub/prj.conf
Comment (de-nordic) [reply tests/drivers/flash_simulator_stub/prj.conf]: Do not understand the comment.

File: tests/drivers/flash_simulator_stub/prj.conf
Comment (de-nordic) [reply tests/drivers/flash_simulator_stub/prj.conf]: I do not trust neither Kconfig nor CMake, so I rather set that to n.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (de-nordic) [reply drivers/flash/flash_simulator_stub/Kconfig]: This can probably be just removed and made always enabled.

**Comment by de-nordic** [comment]
> 
> 
> > Update: Missed the "and then make it possible to use on both sides". OK will do other PR with that proposal too.
> 
> I think the intention was that one driver could be used as either full ram flash mapping, or as a stub (address handling) as it does in current PR state.

The "intention" would require undesired change in upstream. This thing can be removed after mcumgr gets proper support of unpaired slots.

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (de-nordic) [reply drivers/flash/flash_simulator_stub/Kconfig]: The name has been chosen to "FLASH_NOP_DEVICE"

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (de-nordic) [reply drivers/flash/flash_simulator_stub/Kconfig]: > 
> 
> yes, but the DFU is only enabled if mcuboot is enabled, afaik.
> 
> Meaning this symbol is not something a user / conf file should set, but should be automatically selected when mcuboot and DFU for dual image update is enabled.
> And this is being enabled in the app.

You mean that you want to hide the Kconfig options completely?

File: tests/drivers/flash_simulator_stub/prj.conf
Comment (tejlmand) [reply tests/drivers/flash_simulator_stub/prj.conf]: > Do not understand the comment.

the comment is following this comment: https://github.com/nrfconnect/sdk-nrf/pull/5620#discussion_r709154941

So that if the setting is promptless, and defaulting to `n`, then elsewhere in the Kconfig tree we can enable it conditionally by using:
```
if <SOMETHING>

config FLASH_RAM
    bool
    default y

config FLASH_RAM_UNALIGNED_READ
    bool
    default y

endif
```

that `<SOMETHING>` could be the DFU enabling, or mcuboot on thingy53, or something else.
That way it can be guaranteed that requirements to the flash simulator stub is met when a given configuration is enabled.

Relying on settings that has strong dependencies on each other in prj.conf can be fragile if there are no other checks in place.

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:155]: PROG_UNIT ?
ERASE_UNIT affects only erase part, right?

Comment (de-nordic) [reply tests/drivers/flash_nop_device/src/main.c:155]: Sorry, yes. You are right.

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:173]: Here we should check for FLASH_NOP_DEVICE_ERASE_UNIT

File: tests/subsys/dfu/dfu_target_mcuboot/pm.yml
Comment (MarekPieta) [inline tests/subsys/dfu/dfu_target_mcuboot/pm.yml:1]: Is it part of the change ?

**Review by pdunaj — APPROVED** [review]

File: tests/drivers/flash_nop_device/src/main.c
Comment (pdunaj) [reply tests/drivers/flash_nop_device/src/main.c:53]: ...

File: tests/subsys/dfu/dfu_target_mcuboot/pm.yml
Comment (de-nordic) [reply tests/subsys/dfu/dfu_target_mcuboot/pm.yml]: Remved

File: tests/drivers/flash_nop_device/src/main.c
Comment (de-nordic) [reply tests/drivers/flash_nop_device/src/main.c:173]: We are

File: tests/drivers/flash_nop_device/src/main.c
Comment (de-nordic) [reply tests/drivers/flash_nop_device/src/main.c]: Done

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [inline tests/drivers/flash_nop_device/src/main.c:53]: offset is %x and value %02x

File: tests/drivers/flash_nop_device/src/main.c
Comment (MarekPieta) [reply tests/drivers/flash_nop_device/src/main.c:173]: We check for the prog_unit, but since flash_simulator has a hidden dependence between these two we may leave it as it is

**Review by MarekPieta — APPROVED** [review]

File: tests/drivers/flash_nop_device/src/main.c
Comment (pdunaj) [reply tests/drivers/flash_nop_device/src/main.c]: in assert text... in test that will not help anybody now or ever as it is added for thing that returns 0 and writes 0xff. but anyway is fixed I see :)

File: tests/drivers/flash_nop_device/prj.conf
Comment (tejlmand) [inline tests/drivers/flash_nop_device/prj.conf:9]: minor nit:
```suggestion
```

**Review by tejlmand — APPROVED** [review]
lgtm, one minor nit observed, but not blocking in any way as it is in test code.

**Comment by tejlmand** [comment]
@de-nordic thanks for the effort

File: tests/drivers/flash_simulator_stub/prj.conf
Comment (tejlmand) [reply tests/drivers/flash_simulator_stub/prj.conf]: accepted.

File: tests/drivers/flash_simulator_stub/prj.conf
Comment (tejlmand) [reply tests/drivers/flash_simulator_stub/prj.conf]: fixed

File: drivers/flash/flash_simulator_stub/Kconfig
Comment (tejlmand) [reply drivers/flash/flash_simulator_stub/Kconfig]: fixed

**Comment by de-nordic** [comment]
> 
> 
> @de-nordic thanks for the effort

@tejlmand Thanks for helping with Kconfigs.

**Comment by pdunaj** [comment]
There was no manifest change and CI is running for 1 hour...

**Review by pdunaj — CHANGES_REQUESTED** [review]
I heard it's broken. Don't merge before confirmed.

**Review by pdunaj — APPROVED** [review]

**Comment by tejlmand** [comment]
Don't revert an approved PR to an un-approved PR like this:
https://github.com/nrfconnect/sdk-nrf/compare/1f2394b09961bed99ee468b2e9fe18d5d1a9e62e..a1a386b517fda4f14c689dc90bbeb0bc55e8209f

and then merge.
Things are still broken, but not visible, so what my recommendation did, is apparently highlighting dependency chain.

@mbolivar-nordic this should not have been merge, but it's not your fault.

**Comment by tejlmand** [comment]
@pdunaj and @de-nordic don't revert an approved PR without consulting the reviewers.

**Comment by pdunaj** [comment]
> don't revert an approved PR without consulting the reviewers.

Please don't blame @de-nordic for that action as I asked him and everybody else to go this path.
I tried to inform you asap that this is done so we could review the state on Monday. The intention was not to override/ignore anybody but to speed up release testing.

Sometimes there is no other choice. If we had not reverted back to working version we would have nothing today. That would meant we would only be able to merge stub fix by the end of Monday and start testing DFU feature on Tuesday. All that if we were lucky.
The revert was done on assumption that the previous version was tested and working. If on the other hand there were any functional problem with the PR we would be in the same situation - having to fix everything on Monday.

**Comment by tejlmand** [comment]
> I tried to inform you asap that this is done so we could review the state on Monday. The intention was not to override/ignore anybody but to speed up release testing.

I understand your reasons, but if allow such process regardless of the reason, then we completely destroy the review process.
This is identical to authors adding new commits to an already approved review and not requesting a re-review from reviewers.

**Pull request merged**
